### ISSUE TITLE
Cursor Lock Does Nothing on MacOS

### ISSUE THREAD

**Issue by Rigidity** [bevyengine/bevy#4875]
## Bevy version

0.7.0

## Relevant system information

MacOS Macbook Pro 13" running on M1 processor

## What you did

```rs
fn lock_mouse(window: &mut Window) {
    window.set_cursor_visibility(false);
    window.set_cursor_lock_mode(true);

    window.set_cursor_position(Vec2::new(window.width() / 2.0, window.height() / 2.0));
}
```

## What went wrong

- what were you expecting?
For this code to do both correctly, like it does on Windows.

- what actually happened?
On MacOS, this hides the cursor while it's in the window, but does not lock the cursor to the window.

## Additional information

This function is being called on a startup system. Doing it after, in a main system, doesn't work either.

**Comment by alice-i-cecile** [comment]
This is likely an upstream bug in `winit`. We should investigate and fix it there.

**Comment by x-52** [comment]
`Window::set_cursor_lock_mode` calls [`winit::Window::set_cursor_grab`](https://docs.rs/winit/latest/winit/window/struct.Window.html#method.set_cursor_grab). This function doesn't actually lock the cursor to the window—it instead “locks the cursor in a fixed location”, according to the documentation. 

In fact, the [`NSCursor`](https://developer.apple.com/documentation/appkit/nscursor) class (the native API for interacting with the cursor) doesn't even have a `lock` method. 

`winit` simply manually overrides the cursor's position over and over again to keep it at the same spot. There aren't any other options. 

In other words, it's impossible to truly lock a cursor on macOS. But if you keep the app in focus, [the cursor should stay invisible, even if the cursor travels outside the bounds of the window itself](https://docs.rs/winit/latest/winit/window/struct.Window.html#platform-specific-28).

**Comment by zyxkad** [comment]
> In other words, it's impossible to truly lock a cursor on macOS. But if you keep the app in focus, [the cursor should stay invisible, even if the cursor travels outside the bounds of the window itself](https://docs.rs/winit/latest/winit/window/struct.Window.html#platform-specific-28).

On *MacOS*, the cursor will be visible if you move the cursor to the screen border _(even if you are in the full screen mode)_

**Comment by mockersf** [comment]
winit unlocks the cursor when setting the position: 
https://github.com/rust-windowing/winit/blob/1886949efea9f7f76c4cd164a25d8a72ef75e9ea/src/platform_impl/macos/window.rs#L767-L782 (`CGDisplay::associate_mouse_and_mouse_cursor_position(true)` unlocks the cursor)

Setting the position before locking the cursor works as you expect 👍 

If anyone can confirm how it works on other platforms, we could fix it in winit
